### PR TITLE
Week4 Submissions

### PR DESCRIPTION
**Pull request by Evan-CSE**
Week 4 submission 
Topic: Linked List

### CODE REVIEW

File: week03/readme.md [hunk 1]
```diff
@@ -1,3 +1 @@
-Add your week-03's solution to this directory.
-Create a seperate file for each of your solution. Name the file according to the problem name.
```
Comment (Sadman007) [inline week03/readme.md:2]: Please do not touch anything from other weeks.

File: week04/linked_list_cycle.cpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
+/**
+ * TC: O(n)
+ * MC:O(1)
```
Comment (Sadman007) [inline week04/linked_list_cycle.cpp:3]:
```suggestion
 * MC: O(1)
```

File: week04/linked_list_cycle.cpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ */
+class Solution
+{
+public:
+    bool hasCycle(ListNode *head)
+    {
+        ListNode *fast;
+        ListNode *slow;
+        fast = head;
+        slow = head;
```
Comment (Sadman007) [inline week04/linked_list_cycle.cpp:14]:
```suggestion
        ListNode *fast = head;
        ListNode *slow = head;
```

File: week04/maximum_twin_sum_in_a_linked_list.cpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        auto fast = head;
+        ListNode *dummy_node = new ListNode(-1, head);
+        auto slow = dummy_node;
+        int max_sum = 0;
+        while (fast && fast->next)
+        {
+            fast = fast->next->next;
+            slow = slow->next;
+        }
+        // Now slow is in the middle of the list
```
Comment (Sadman007) [inline week04/maximum_twin_sum_in_a_linked_list.cpp:33]:
```suggestion
        // Now `slow` is in the middle of the list
        // now reverse the list from this point
```

File: week04/maximum_twin_sum_in_a_linked_list.cpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        // now reverese the list from this point
```
Comment (Sadman007) [inline week04/maximum_twin_sum_in_a_linked_list.cpp:34]:
```suggestion
        // Now slow is in the middle of the list
        // now reverse the list from this point
```

File: week04/merge_k_sorted_list.cpp [hunk 6]
```diff
@@ -0,0 +1,55 @@
+/*
+    TC: O(k*n)
+    where,
+         k is the length of the vector of lists
+         n is the total number of elements in the final list
```
Comment (Sadman007) [inline week04/merge_k_sorted_list.cpp:5]: Wrong.

File: week04/populating_next_right_pointer.cpp [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        queue<Node *> q;
+        q.push(root);
+        int processed = 0;
+        while (!q.empty())
+        {
+            Node *current_node = q.front();
+            processed++;
+            q.pop();
+            if (isPowerOfTwo(processed + 1))
+                current_node->next = NULL;
```
Comment (Sadman007) [inline week04/populating_next_right_pointer.cpp:29]: This logic could be omitted and it's redundant.

File: week04/remove_nth_node_from_end_of_the_list.cpp [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+{
+public:
+    ListNode *removeNthFromEnd(ListNode *head, int n)
+    {
+        auto fast = head;
+        ListNode *dummy_node = new ListNode(0, head);
+        auto slow = dummy_node;
+        int displacement = 0;
+        while (displacement < n)
+            fast = fast->next, displacement++;
```
Comment (Sadman007) [inline week04/remove_nth_node_from_end_of_the_list.cpp:16]: Great

File: week04/reverse_linked_list.cpp [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+ * TC: O(n)
+ * MC:O(1)
+ * where n is the number of elements in the list
+ */
+class Solution
+{
+public:
+    ListNode *reverseList(ListNode *cur)
+    {
+        ListNode *nxt = new ListNode(0, NULL);
```
Comment (Sadman007) [inline week04/reverse_linked_list.cpp:11]: Better define what 0 is. It's better to declare a static dummy value.

**Review by Sadman007 — APPROVED** [review]
Hi @Evan-CSE 

Feel free to merge the PR after **resolving the comments (if any)**. Thanks.

**Comment by Evan-CSE** [comment]
Thanks for your valuable feedback, vai. I have resolved all issues except 'merge_k_sorted_list'. I need some time to solve this. But I have commented out this suggestion in my code so that I can optimise the code.

**Pull request closed without merging**